src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -4,18 +4,279 @@
+        /// The <see cref="ClrThreadPool.WaitThread"/> this <see cref="RegisteredWaitHandle"/> was registered on.
+        /// </summary>
+        internal ClrThreadPool.WaitThread WaitThread { get; set; }
+
+        /// <summary>
+        /// The number of callbacks that are currently queued on the Thread Pool or executing.
+        /// </summary>
+        private int _numRequestedCallbacks;
+
+        private LowLevelLock _callbackLock = new LowLevelLock();
+
+        /// <summary>
+        /// Notes if we need to signal the user's unregister event after all callbacks complete.
+        /// </summary>
+        private bool _signalAfterCallbacksComplete;
+
+        private int _unregisterCalled;
+
+        private bool _unregistered;
+
+        private AutoResetEvent _callbacksComplete;
+
+        private AutoResetEvent _removed;
+
+        /// <summary>
+        /// Unregisters this wait handle registration from the wait threads.
+        /// </summary>
+        /// <param name="waitObject">The event to signal when the handle is unregistered.</param>
+        /// <returns>If the handle was successfully marked to be removed and the provided wait handle was set as the user provided event.</returns>
+        /// <remarks>
+        /// This method will only return true on the first call.
+        /// Passing in a wait handle with a value of -1 will result in a blocking wait, where Unregister will not return until the full unregistration is completed.
+        /// </remarks>
         public bool Unregister(WaitHandle waitObject)
         {
-            // UNIXTODO: ThreadPool
-            throw new NotImplementedException();
+            if (Interlocked.Exchange(ref _unregisterCalled, 1) == 0)
+            {
+                GC.SuppressFinalize(this);
+
+                _callbackLock.Acquire();
+                try
+                {
+                    UserUnregisterWaitHandle = waitObject?.SafeWaitHandle;
+                    UserUnregisterWaitHandle?.DangerousAddRef();
+                    UserUnregisterWaitHandleValue = UserUnregisterWaitHandle?.DangerousGetHandle() ?? IntPtr.Zero;
+                    if (_unregistered)
+                    {
+                        SignalUserWaitHandle();
+                        return true;
+                    }
+
+                    _removed = new AutoResetEvent(false);
+
+                    if (IsBlocking)
+                    {
+                        _callbacksComplete = new AutoResetEvent(false);
+                    }
+                }
+                finally
+                {
+                    _callbackLock.Release();
+                }
+
+                WaitThread.UnregisterWait(this);
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Signal <see cref="UserUnregisterWaitHandle"/> if it has not been signaled yet and is a valid handle.
+        /// </summary>
+        private void SignalUserWaitHandle()
+        {
```

Comment:
Please add a `_callbackLock.VerifyIsLocked();` at the top so that it's immediately clear when reading that this function is only called under the lock